typescript/web/src/components/import-button/import-images-modal/modal-dropzone/import-dropped-files/import-datasets.tsx

```diff
@@ -0,0 +1,95 @@
+import { ApolloClient, gql } from "@apollo/client";
+import { ExportFormat } from "@labelflow/graphql-types";
+import { v4 as uuidv4 } from "uuid";
+import mime from "mime-types";
+
+import Bluebird from "bluebird";
+import { uploadFile } from "../../../../../utils/upload-file";
+import { DroppedFile, SetUploadStatuses } from "../../types";
+
+import { CONCURRENCY } from "../../constants";
+
+const importDatasetMutation = gql`
+  mutation importDataset(
+    $where: DatasetWhereUniqueInput!
+    $data: DatasetImportInput!
+  ) {
+    importDataset(where: $where, data: $data) {
+      error
+    }
+  }
+`;
+const importDataset = async ({
+  apolloClient,
+  datasetId,
+  workspaceId,
+  file,
+}: {
+  apolloClient: ApolloClient<any>;
+  datasetId: string;
+  workspaceId: string;
+  file: Blob;
+}) => {
+  const key = `${workspaceId}/${datasetId}/import-datasets/${uuidv4()}-${new Date().toISOString()}.${mime.extension(
+    file.type
+  )}`;
+  const url = await uploadFile({ file, key, apolloClient });
+
+  const dataImportDataset = await apolloClient.mutate({
+    mutation: importDatasetMutation,
+    variables: {
+      where: { id: datasetId },
+      data: {
+        url,
+        format: ExportFormat.Coco,
+        options: {
+          coco: {
+            annotationsOnly: file.type === "application/json",
+          },
+        },
+      },
+    },
+    refetchQueries: [
+      "getDatasetData",
+      "getImageLabels",
+      "getLabel",
+      "countLabelsOfDataset",
+      "getDatasetLabelClasses",
+    ],
+  });
+  if (dataImportDataset?.data?.importDataset?.error) {
+    throw new Error(dataImportDataset?.data?.importDataset?.error);
+  }
+};
+
+export const importDatasets = async ({
+  datasets,
+  datasetId,
+  workspaceId,
+  apolloClient,
+  setFileUploadStatuses,
+}: {
+  datasets: DroppedFile[];
+  datasetId: string;
+  workspaceId: string;
+  apolloClient: ApolloClient<any>;
+  setFileUploadStatuses: SetUploadStatuses;
+}) => {
+  return await Bluebird.Promise.map(
```

Comment:
Why do you use Bluebird?

Reply:
That's the standard library we used until now when we want to map on an array with a limited concurrency.